lxd/auth/driver_openfga.go

```diff
@@ -0,0 +1,1079 @@
+		Relation: relationProject,
+		Type:     string(ObjectTypeImage),
+	}).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	imageObjects := resp.GetObjects()
+	allObjects = append(allObjects, imageObjects...)
+
+	resp, err = f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
+		User:     ObjectProject(projectName).String(),
+		Relation: relationProject,
+		Type:     string(ObjectTypeImageAlias),
+	}).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	imageAliasObjects := resp.GetObjects()
+	allObjects = append(allObjects, imageAliasObjects...)
+
+	resp, err = f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
+		User:     ObjectProject(projectName).String(),
+		Relation: relationProject,
+		Type:     string(ObjectTypeNetwork),
+	}).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	networkObjects := resp.GetObjects()
+	allObjects = append(allObjects, networkObjects...)
+
+	resp, err = f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
+		User:     ObjectProject(projectName).String(),
+		Relation: relationProject,
+		Type:     string(ObjectTypeNetworkACL),
+	}).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	networkACLObjects := resp.GetObjects()
+	allObjects = append(allObjects, networkACLObjects...)
+
+	resp, err = f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
+		User:     ObjectProject(projectName).String(),
+		Relation: relationProject,
+		Type:     string(ObjectTypeNetworkZone),
+	}).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	networkZoneObjects := resp.GetObjects()
+	allObjects = append(allObjects, networkZoneObjects...)
+
+	resp, err = f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
+		User:     ObjectProject(projectName).String(),
+		Relation: relationProject,
+		Type:     string(ObjectTypeProfile),
+	}).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	profileObjects := resp.GetObjects()
+	allObjects = append(allObjects, profileObjects...)
+
+	resp, err = f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
+		User:     ObjectProject(projectName).String(),
+		Relation: relationProject,
+		Type:     string(ObjectTypeStorageVolume),
+	}).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	storagePoolVolumeObjects := resp.GetObjects()
+	allObjects = append(allObjects, storagePoolVolumeObjects...)
+
+	resp, err = f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
+		User:     ObjectProject(projectName).String(),
+		Relation: relationProject,
+		Type:     string(ObjectTypeStorageBucket),
+	}).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	storageBucketObjects := resp.GetObjects()
+	allObjects = append(allObjects, storageBucketObjects...)
+	return allObjects, nil
+}
+
+func (f *fga) syncResources(ctx context.Context, resources Resources) error {
```

Comment:
Same thing here: could we use loops and common pattern and abstract repetitive tasks into helper functions where possible ? For example:

```golang
func (f *fga) syncResources(ctx context.Context, resources Resources) error {
	var writes, deletions []client.ClientTupleKey

	options := client.ClientListObjectsOptions{AuthorizationModelId: openfga.PtrString(f.authModelID)}
	serverUserStr := ObjectServer().String()

	// Helper function to sync resources
	sync := func(resourceType ObjectType, currentResources, existingResources []Object) error {
		for _, currResource := range currentResources {
			if !shared.ValueInSlice(currResource.String(), existingResources) {
				writes = append(writes, client.ClientTupleKey{
					User:     serverUserStr,
					Relation: relationServer,
					Object:   currResource.String(),
				})
			}
		}

		for _, existingResource := range existingResources {
			resourceObject, err := ObjectFromString(existingResource)
			if err != nil {
				return err
			}

			if !shared.ValueInSlice(resourceObject, currentResources) {
				deletions = append(deletions, client.ClientTupleKey{
					User:     serverUserStr,
					Relation: relationServer,
					Object:   existingResource,
				})
			}
		}

		return nil
	}

	resourceMappings := []struct {
		Type       ObjectType
		Resources  []Object
	}{
		{ObjectTypeCertificate, resources.CertificateObjects},
		{ObjectTypeStoragePool, resources.StoragePoolObjects},
		{ObjectTypeProject, resources.ProjectObjects},
	}

	for _, mapping := range resourceMappings {
		resp, err := f.client.ListObjects(ctx).Options(options).Body(client.ClientListObjectsRequest{
			User:     serverUserStr,
			Relation: relationServer,
			Type:     string(mapping.Type),
		}).Execute()
		if err != nil {
			return err
		}

		err = sync(mapping.Type, mapping.Resources, resp.GetObjects())
		if err != nil {
			return err
		}
	}

	// Special handling for project resources
	var existingProjectResources []string
	for _, existingProject := range existingProjects {  // Assuming `existingProjects` is fetched somewhere before this line
		projectObject, err := ObjectFromString(existingProject)
		if err != nil {
			return err
		}

		projectResources, err := f.projectObjects(ctx, projectObject.Project())
		if err != nil {
			return err
		}

		existingProjectResources = append(existingProjectResources, projectResources...)
	}

	currentProjectResources := append(resources.ImageObjects, resources.ImageAliasObjects...)
	currentProjectResources = append(currentProjectResources, resources.InstanceObjects...)
	// ... keep appending the other resources
	err := sync(ObjectTypeProject, currentProjectResources, existingProjectResources)
	if err != nil {
		return err
	}

	return f.updateTuples(ctx, writes, deletions)
}
```

Reply:
This is really good feedback, thank you :) I'll work on refining the drivers to remove any repeated logic.